### PR TITLE
[Feature] Merge schema in `FILES` schema detection

### LINKED ISSUE #36961: Merge schema in `files` schema detection
## Enhancement
The schema of files may be different since the source has been altered by DDL statement.
It's necessary to merge schema in `files` schema detection, to make it possible to load files in different schema via `files`.

### PR DESCRIPTION
**Pull request by rickif**
Why I'm doing:
Now, the `files` only sample 1 file of the files list to get the schema. When the files are in different schemas, they cannot be loaded by `files`.

What I'm doing:
This PR adds the schema merging to the schema detection process of `files`. `auto_detect_sample_files` is added as a property to control the number of sample files.
```
MySQL [db0]> select * from files("path"="hdfs://127.0.0.1:9000/merge-schema1/*", "format"="parquet", "auto_detect_sample_files"="2");
+------+------+
| col1 | col3 |
+------+------+
| 5    | 7    |
| 6    | 8    |
| 9    | 11   |
| 10   | 12   |
+------+------+
4 rows in set (0.036 sec)
```

The schema merging rules are:
1. When the input file columns are different, the output is the total of all columns
2. The column types from different files are different.
2.1 If both types are in integer, the final output column is of type with promoted type. eg: `TYPE_INT`, `TYPE_BIGINT`-> `TYPE_BIGINT`.
2.2 if both types are in float, the final output column is of type with `TYPE_DOUBLE`. eg: `TYPE_FLOAT`, `TYPE_DOUBLE`-> `TYPE_DOUBLE`.
2.3 Otherwise, the final output column is of type `TYPE_VARCHAR`. eg: `TYPE_FLOAT`, `TYPE_INT` -> `TYPE_VARCHAR`.
3. Sampling files were evenly selected from the file list according to the configuration `auto_detect_sample_files`
4. If there're same column names in different cases (eg: col1, COL1), an error `Identical names in upper/lower cases, file:  [/work/starrocks-main/be/test/exec/test_data/parquet_data/schema4.parquet], column names: [col1 [COL1]` would be returned.
Fixes #36961 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by Kimi-SH** [comment]
I've received your email and will will reply to you as soon as possible after the vacation.————————————This is an auto reply email from QQ Mail.

**Comment by rickif** [comment]
@Metgifyio rebase

**Comment by LittleBeeBee** [comment]
您好，您的邮件已收到，我会尽快给你回复。

**Comment by kobebryantlin0** [comment]
您好，我是林恩伟，你发的邮件我已经收到。

**Comment by rickif** [comment]
@mergifyio rebase

**Review by wyb — DISMISSED** [review]

File: be/src/exec/file_scanner.cpp [hunk 1]
```diff
@@ -355,4 +358,109 @@ Status FileScanner::create_random_access_file(const TBrokerRangeDesc& range_desc
...
+void merge_schema(const std::vector<std::vector<SlotDescriptor>>& input, std::vector<SlotDescriptor>* output) {
+    if (output == nullptr) {
+        return;
+    }
+
+    std::vector<std::shared_ptr<SlotDescriptor>> merged_schema;
+    std::map<std::string, size_t> merged_schema_index;
+    for (const auto& schema : input) {
+        for (const auto& slot : schema) {
+            auto itr = merged_schema_index.find(slot.col_name());
```
Comment (wyb) [inline be/src/exec/file_scanner.cpp:370]: deal with the case of column name? sensitive or insensitive？

Comment (rickif) [reply be/src/exec/file_scanner.cpp:370]: It's case-senstitive now.

**Review by wyb — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Comment by rickif** [comment]
@mergifyio backport branch-3.2

**Pull request merged**